README.md
<div align="center">
  <p>
 <a href="https://www.npmjs.com/package/qmy"><img  src="https://raw.githubusercontent.com/4i8/qmy/main/logo/Qmy.png" width="400" alt="qmy" /></a>
  </p>
  <p>
 <a href="https://github.com/4i8/"><img src="https://img.shields.io/static/v1?label=powered%20by&message=Arth&color=000&style=for-the-badge&logo=Windows%20Terminal&logoColor=fff"/></a>
 <a href="https://www.npmjs.com/package/qmy"><img src="https://img.shields.io/npm/v/qmy.svg?style=for-the-badge" alt="NPM version" /></a>
 <a href="https://www.npmjs.com/package/qmy"><img src="https://img.shields.io/npm/dt/qmy.svg?maxAge=3600&style=for-the-badge" alt="NPM downloads" /></a>
 <a href="https://discord.gg/r7sgerWCcT"><img src="https://img.shields.io/discord/894463244188676116?https://img.shields.io/static/v1?text=f&style=for-the-badge&logo=discord&logoColor=fff" alt="Discord server" /></a>

  </p>
</div>

## About

Quick [mysql](https://www.mysql.com/) is a powerful module that makes the mysql system easier, simpler and faster

- Speed
- Unlimited connection without losing connection
- Easy to use
- High security

## Installation

```sh-session
npm install qmy
yarn add qmy
```

## Announcement 

> **(+)connect() will resolve the process by calling the then() function and it will give you a notification that the connection was successful if there is an error, it will reject the process and you can call the catch() function and it will give you the error<br>(-)Remove Reminders New Versions**

## Example usage

> **Connection Config**

```js
 host: "localhost",
 port: 3306,
 localAddress: undefined,
 socketPath: undefined,
 user: undefined,
 password: undefined,
 database: undefined,
 connectTimeout: 10000,
 insecureAuth: false,
 supportBigNumbers: false,
 bigNumberStrings: false,
 dateStrings: false,
 debug: undefined,
 trace: true,
 stringifyObjects: false,
 timezone: "local",
 flags: "",
 queryFormat: undefined,
 pool: undefined,
 ssl: false,
 localInfile: true,
 multipleStatements: false,
 typeCast: true,
 maxPacketSize: 0,
 charsetNumber: 33,
 clientFlags: 455631,
 protocol41: true,
```

### #Note

##### Direct connection problem solving: [Here](#connect)

> **create connection**

**Start by creating a connection to the database.
Use the username and password from your MySQL database or use uri**

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
```

## Documentation

<table>
  <tr>
    <th>Method</th>
    <th>Description</th>
    <th>explain</th>
  <tr>
    <td>connect</td>
    <td>connect() will resolve the process by calling the then() function and it will give you a notification that the connection was successful if there is an error, it will reject the process and you can call the catch() function and it will give you the error
</td>
    <td><a href=#connect>Go to example</a></td>
  </tr>
   <tr>
    <td>query</td>
    <td>Use SQL statements to read from (or write to) a MySQL database. This is also called "to query" the database.
The connection object created in the example above, has a method for querying the database</td>
    <td><a href=#query>Go to example</a></td>
  </tr>
     <tr>
    <td>get</td>
    <td>used to select data from a database</td>
    <td><a href=#get>Go to example</a></td>
  </tr>
    <tr>
    <td>remove</td>
    <td>used to delete existing records in a table</td>
    <td><a href=#remove>Go to example</a></td>
  </tr>
    <tr>
    <td>set</td>
    <td>used to insert new records in a table</td>
    <td><a href=#set>Go to example</a></td>
  </tr>
   </tr>
    <tr>
    <td>update</td>
    <td>used to modify the existing records in a table</td>
    <td><a href=#update>Go to example</a></td>
  </tr>
</table>

## connect

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    //do something
    //...
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## query

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    connection
      .query("<SQL statements>")
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## get

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    connection
      .get("<table>", "<Where>:<PrimaryKey>")
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## remove

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    connection
      .remove("<table>", "<Where>:<PrimaryKey>")
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## set

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    connection
      .set(
        {
          table: "users",
          column: "id - name - color",
          values: "2324249073 - arth - red",
        },
        { sign: "-" }
      )
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## update

```js
const qmy = require("qmy");

const connection = new qmy({
  database: "<database>",
  host: "<host>",
  password: "<password>",
  port: 3306,
  user: "<user>",
});
connection
  .connect()
  .then(() => {
    connection
      .update({
        table: "users",
        column: "name",
        PrimaryKey: "id:2324249073",//<Where>:<PrimaryKey>
        value: "arth",
      })
      .then((result) => {
        console.log(result);
      })
      .catch((err) => {
        console.log(err);
      });
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });
```

## Links

- [Twiter](https://twitter.com/onlyarth)
- [Github](https://github.com/4i8)
- [Documentation](#documentation)

## License

- [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0)


Changes.md
<div align="center">
  <p>
 <a href="https://www.npmjs.com/package/qmy"><img  src="https://raw.githubusercontent.com/4i8/qmy/main/logo/Qmy.png" width="400" alt="qmy" /></a>
  </p>
  <p>
 <a href="https://github.com/4i8/"><img src="https://img.shields.io/static/v1?label=powered%20by&message=Arth&color=000&style=for-the-badge&logo=Windows%20Terminal&logoColor=fff"/></a>
 <a href="https://www.npmjs.com/package/qmy"><img src="https://img.shields.io/npm/v/qmy.svg?style=for-the-badge" alt="NPM version" /></a>
 <a href="https://www.npmjs.com/package/qmy"><img src="https://img.shields.io/npm/dt/qmy.svg?maxAge=3600&style=for-the-badge" alt="NPM downloads" /></a>
 <a href="https://discord.gg/r7sgerWCcT"><img src="https://img.shields.io/discord/894463244188676116?https://img.shields.io/static/v1?text=f&style=for-the-badge&logo=discord&logoColor=fff" alt="Discord server" /></a>

  </p>
</div>

#Changes:

- **[03 / 10 / 2021]**:
  > **Fix a minor bug in main code**
- **[04 / 10 / 2021]**:
  > **A reminder was added to add a new version**
- **[9 / 10 / 2021]**:
  > **Fix a minor bug in main code**
- **[24 / 10 / 2021]**:
  > **1-The disconnection issue has been fixed
  > 2-The options issue has been fixed
  > 3-The log file issue has been fixed
  > 4-Internal issues in Functions have been fixed
  > 5-Connection issues fixed
  > 6-Internal issues with file have been fixed
  > 7-Package files have been rearranged
  > 8-The log files have been rearranged
  > 9-Some internal features have been added
  > 10-A new option has been added**
- **[5 / 12 / 2021]**:
  > **1-All mySQL options have been added, you can also put a uri to connect to MYSQL
  > 2-The LogFile option has been removed
  > 3-Callback was replaced to promise(then and catch and finally)
  > 4-Internal issues in Functions have been fixed
  > 5-The set function have been made now you can place a custom Sign
  > 6-Internal issues with file have been fixed
  > 7-result of get function be object and you have a option you can make Array
  > 8-result of query function be Array and you have a option you can make object**
- **[21 / 02 / 2022]**: 
  > **(+)connect() will resolve the process by calling the then() function and it will give you a notification that the connection was successful if there is an error, it will reject the process and you can call the catch() function and it will give you the error<br>(-)Remove Reminders New Versions**


index.js
/**
 * @Copyright 2021 Arth(https://github.com/4i8/)
 */
const mysql = require("mysql");
const got = require("got");
const v = require("./package.json");
const events = require("events");
//METHODS
const get = require("./src/methods/get");
const set = require("./src/methods/set");
const update = require("./src/methods/update");
const remove = require("./src/methods/remove");
const query = require("./src/methods/query");
const custom = require("./src/methods/log/custom");
const parseUrl = require("./src/methods/other/parseUrl");
const EventEmitter = new events();
/**
 * Create a new Connection 
 ```js
 (connectionUri: string | ConnectionConfig): mysql.Connection
 ```
 */
function connection(
  ConnectionConfig = {
    host: "localhost",
    port: 3306,
    localAddress: undefined,
    socketPath: undefined,
    user: undefined,
    password: undefined,
    database: undefined,
    connectTimeout: 10000,
    insecureAuth: false,
    supportBigNumbers: false,
    bigNumberStrings: false,
    dateStrings: false,
    debug: undefined,
    trace: true,
    stringifyObjects: false,
    timezone: "local",
    flags: "",
    queryFormat: undefined,
    pool: undefined,
    ssl: false,
    localInfile: true,
    multipleStatements: false,
    typeCast: true,
    maxPacketSize: 0,
    charsetNumber: 33,
    clientFlags: 455631,
    protocol41: true,
  },
  options = {
    connection: {
      log: (connectionlog = true),
    },
  }
) {
  try {
    var database = ConnectionConfig.database;
    if (typeof ConnectionConfig === "string") {
      database = parseUrl(ConnectionConfig).database;
    }
    var connectionlog = options.connection.log;
  } catch {}
  const createCon = () => {
    let db = mysql.createConnection(ConnectionConfig);
    return db;
  };
  const reConnect = (connection, error) =>
    new Promise((resolve) => {
      if (!error.toString().includes("Packets out of order")) {
        setTimeout(() => {
          EventEmitter.emit("error", error);
        }, 1);
        throw error;
      }
      connection.destroy();
      if (connectionlog == true)
        custom("MYSQL Connection is Restarted", "F5E21E");
      setTimeout(() => resolve(), 3500);
    });

  const Connect = () =>
    new Promise((resolve) => {
      const con = createCon();
      con.connect(async (err) => {
        if (err) {
          if (!err.code.toString().includes("Packets out of order")) {
            setTimeout(() => {
              EventEmitter.emit("error", err);
            }, 1);
          }
          try {
            setTimeout(() => {
              EventEmitter.emit("error", err);
            }, 1);
          } catch (e) {
            await reConnect(con, e);
            Connect();
          }
        } else {
          if (connectionlog == true)
            custom("Connect To MYSQL Successfully", "1EF59D");
          setTimeout(() => {
            EventEmitter.emit("connect", "Connect To MYSQL Successfully");
          }, 1);
          return resolve(con);
        }
      });
    });
  var Connection;
  const connection = () =>
    new Promise(async (resolve) => {
      Connection = await Connect();
      Connection.on("error", async (e) => {
        await reConnect(Connection, e);
        connection();
      });
      return resolve();
    });
  (async () => await connection())();
  //<- START ->
  // <-connect->
  /**
   *connect() will resolve the process by calling the then() function and it will give you a notification that the connection was successful if there is an error, it will reject the process and you can call the catch() function and it will give you the error
* @example 
connect()
  .then(() => {
    //do something
    //...
  })
  .catch((error) => {
    //do something
    //...
    console.log(error);
  });

   */
  this.connect = function () {
    return new Promise((resolve, reject) => {
      EventEmitter.on("connect", () => {
        resolve();
      });
      EventEmitter.on("error", (error) => {
        reject(error);
      });
    });
  };
  // <-query->
  /**
*Use SQL statements to read from (or write to) a MySQL database. This is also called "to query" the database. The connection object created in the example above, has a method for querying the database
* @param  {array: <default:true>:boolean }
* @example query("<SQL statements>").then((result) => {
    console.log(result)
}).catch((err) => {
    console.log(err);
  });
*/
  this.query = function (sql, options = { array: true }) {
    return query(sql, Connection, options);
  };
  // <-get->
  /**
*	used to select data from a database
* @param  {array: <default:false>:boolean }
* @example get("<table>", "<Where>:<PrimaryKey>").then((result) => {
    console.log(result);
  }).catch((err) => {
    console.log(err);
  });
* Example of a <Where>:<PrimaryKey>
 @example get("users", "id:2324249073").then((result) => {
    console.log(result);
  }).catch((err) => {
    console.log(err);
  });
*/
  this.get = function (table, PrimaryKey, options = { array: false }) {
    return get(table, PrimaryKey, Connection, database, options);
  };
  // <-remove->
  /**
  *used to delete existing records in a table	
  *
  * @example remove("<table>", "<Where>:<PrimaryKey>").then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
  * Example of a <Where>:<PrimaryKey>
 @example 
  remove("users", "id:2324249073").then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
   */
  this.remove = function (table, PrimaryKey) {
    return remove(table, PrimaryKey, Connection, database, options);
  };
  // <-set->
  /**
   *Tused to insert new records in a table	
   * @param  {sign: <default:",">:string }
   * @example set({
    "table": "<table>", 
    "column": "<column>",
     "values": "<values>"
    }).then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
   //It is placed sign if there is more than one value or column
   @example set({
    "table": "users", 
    "column": "id - name - color",
    "values": "2324249073 - arth - red"}, { sign:"-" }).then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
   */
  this.set = function (
    { table: table, column: column, values: values },
    options = { sign: "," }
  ) {
    return set(
      { table: table, column: column, values: values },
      Connection,
      database,
      options
    );
  };
  // <- Data update ->
  /**
   *used to modify the existing records in a table
   *
   * @example update({
       table: "<table>",
       column: "<column>",
       PrimaryKey: "<Where>:<PrimaryKey>",
       value: "<value>",
     }).then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
   * Example of a <Where>:<PrimaryKey>
 @example 
 update({
       table: "users",
       column: "name",
       PrimaryKey: "id:2324249073",
       value: "arth",
     }).then((result) => {
    console.log(result)
 }).catch((err) => {
    console.log(err);
  });
   */
  this.update = function ({
    table: table,
    column: column,
    value: value,
    PrimaryKey: PrimaryKey,
  }) {
    return update(
      { table: table, column: column, value: value, PrimaryKey: PrimaryKey },
      Connection,
      database,
      options
    );
  };
//   (async () => {
//     try {
//       const data = await got.post("https://img.shields.io/npm/v/qmy.json", {
//         responseType: "json",
//       });
//       if ("v" + v.version !== `${data.body.value}`) {
//         console.log(
//           `
// # New Version - ${data.body.value}           
// ╔═════════════╦═════════════════════════════════╗              
// ║ Install     ║ npm install qmy                 ║
// ╠═════════════╬═════════════════════════════════╣
// ║ #Note       ║ When you update to the latest   ║
// ║             ║ version, you don't get any bugs ║
// ╚═════════════╩═════════════════════════════════╝
//       `
//         );
//       }
//     } catch (error) {}
//   })();
  //<- END ->
}
module.exports = connection;
/**
 * @Copyright 2021 Arth(https://github.com/4i8/)
 */
